From fe44cd7e988e8480a1a5be4b9e953136d1173da8 Mon Sep 17 00:00:00 2001
From: Ramakrishna Pallala <ramakrishna.pallala@intel.com>
Date: Mon, 2 Mar 2015 23:07:36 +0530
Subject: [PATCH] max17072_battery: Fix battery over voltage issue

This patch fixes the battery over voltage reporting issue
by reading the instantaneous voltage to determine the
over voltage condition.

Change-Id: I222fb65d4426d9a7d6e6ba877fb641dce276ccfd
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-21623
Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>
---
 drivers/power/max17042_battery.c | 29 ++++++++++++++++++++---------
 1 file changed, 20 insertions(+), 9 deletions(-)

diff --git a/drivers/power/max17042_battery.c b/drivers/power/max17042_battery.c
index 18e18de..54f4189 100644
--- a/drivers/power/max17042_battery.c
+++ b/drivers/power/max17042_battery.c
@@ -179,7 +179,7 @@ static int max17042_get_temperature(struct max17042_chip *chip, int *temp)
 
 static int max17042_get_battery_health(struct max17042_chip *chip)
 {
-	int temp, ret;
+	int temp, vavg, vbatt, ret;
 	u32 val;
 
 	/* Cannot judge health of an unknown battery */
@@ -187,20 +187,31 @@ static int max17042_get_battery_health(struct max17042_chip *chip)
 		return POWER_SUPPLY_HEALTH_UNKNOWN;
 
 	ret = regmap_read(chip->regmap, MAX17042_AvgVCELL, &val);
-	if (ret < 0)
+	if (ret < 0) {
 		return POWER_SUPPLY_HEALTH_UNSPEC_FAILURE;
-	else {
+	} else {
 		/* bits [0-3] unused */
-		val *= MAX17042_VOLTAGE_CONV_FCTR / 8;
+		vavg = val * MAX17042_VOLTAGE_CONV_FCTR / 8;
 		/* Convert to milli volts */
-		val /= 1000;
-		if (val < chip->pdata->vmin)
-			return POWER_SUPPLY_HEALTH_DEAD;
+		vavg /= 1000;
+	}
 
-		if (val > chip->pdata->vmax + MAX17042_VMAX_OFFSET)
-			return POWER_SUPPLY_HEALTH_OVERVOLTAGE;
+	ret = regmap_read(chip->regmap, MAX17042_VCELL, &val);
+	if (ret < 0) {
+		return POWER_SUPPLY_HEALTH_UNSPEC_FAILURE;
+	} else {
+		/* bits [0-3] unused */
+		vbatt = val * MAX17042_VOLTAGE_CONV_FCTR / 8;
+		/* Convert to milli volts */
+		vbatt /= 1000;
 	}
 
+	if (vavg < chip->pdata->vmin)
+		return POWER_SUPPLY_HEALTH_DEAD;
+
+	if (vbatt > chip->pdata->vmax + MAX17042_VMAX_OFFSET)
+		return POWER_SUPPLY_HEALTH_OVERVOLTAGE;
+
 	ret = max17042_get_temperature(chip, &temp);
 
 	if (ret < 0)
-- 
1.9.1

